site/content/user-guide/sigfox-adapter.md

```diff
@@ -0,0 +1,126 @@
++++
+title = "Sigfox Adapter"
+weight = 260
++++
+
+The Sigfox protocol adapter exposes an HTTP endpoint for connecting up with
+the Sigfox backend for publishing telemetry, events and use command & control.
+<!--more-->
+
+{{% note title="Tech preview" %}}
+This protocol adapter is not considered production ready. Its APIs might still
+be subject to change without warning.
+{{% /note %}}
+
+## Pre-requisites
+
+This Sigfox adapter only connects to the Sigfox backend system
+(`backend.sigfox.com`). It does not allow direct access to Sigfox devices.
+
+So you need to set up your Sigfox devices on `backend.sigfox.com` and then
+configure the callbacks connect to your installation of Hono.
+
+## Devices and credentials
+
+In a nutshell, the Sigfox adapter requires a single *device* identity, acting
+as a gateway device. This identity will be used to connect to Hono. All devices
+registered with Sigfox (the actual Sigfox devices), will be registered in Hono
+to allow this first identity as their gateway device.
+
+## Setup example
+
+The following sub-sections walk you through an example setup.
+
+### Registering devices
+
+This example assumes that the Sigfox protocol adapter is available as
+`https://iot-sigfox-adapter.my.hono`.
+
+Create a new *gateway device* with the following registration:
+
+~~~json
+{
+  "device-id": "sigfox-backend"
+}
+~~~
+
+Create new credentials for the *gateway device*. For example, using
+a username of `sigfox` and a password of `test12`:
+
+~~~json
+{
+  "auth-id": "sigfox",
+  "device-id": "sigfox-backend",
+  "enabled": true,
+  "secrets": [
+    {
+      "hash-function": "sha-512",
+      "pwd-hash": "yHRiqBGqGlGBctRvzX98JpeTTk0ao9BfzR714G7737y8oOuRsFzJYGC0846lFy96FWev8sSb7MGeSqoJi3I6mw==",
+      "salt": "j9PyA6Y9obw="
+    }
+  ],
+  "type": "hashed-password"
+}
+~~~
+
+Create a new *device*, referencing the previous *gateway device*. The
+*device id* must be your Sigfox device ID (e.g. `1AB2C3`):
+
+~~~json
+{
+  "device-id": "1AB2C3",
+  "via": "sigfox-backend"
+}
+~~~
+
+### Setting up callbacks
+
+Log in to the Sigfox backend at https://backend.sigfox.com and then open up
+the view `Device Type` -> `Callbacks`.
+
+Create a new "Custom" callback, with the following settings:
+
+* **Type**: `DATA` – `UPLINK`
+* **Channel**: `URL`
+* **Url pattern**: `https://iot-sigfox-adapter.my.hono/data/telemetry?device={device}&data={data}`
+* **Use HTTP Method**: `GET`
```

Comment:
can't we PUT the data to `telemetry/{tenant}/{device} ?

Reply:
Sure we could. The default selection in the Sigfox UI is "GET" however. The protocol adapter actually implements all three (GET, PUT, POST) with the same operation. And the proposal is also to use query parameters.

Reply:
Actually it is not possible to send the `data` argument in the request body. So even using PUT, we would need to use the query parameter.

Reply:
I see :-(

Reply:
I have to correct myself, it looks like for `POST` and `PUT` it is possible to transmit this using `application/x-www-form-urlencoded`. Not sure this is an improvement though. I will play around with this a bit and see what is actually possible.

In any case, I think we could improve this to:

    /data/{device}/telemetry?data={data}

Considering other URLs, for other options/callbacks:

    /data/{device}/event?data={data}
    /error/{device}/event?data={data}
    /status/{device}/event?info={info}

I don't think we need the tenant information a second time, as it is already included in the authentication credentials.

Reply:
So it turns you actually can specify in more detail what the "body" section may look like when choosing PUT or POST and putting in the different values for the content type.

However, that would mean you need to put in part of the information (device) in the URL and part into the body. But first remove all the default information from the Web UI.

So I think the most simple approach is to keep the current way of passing this via query parameters.

Reply:
what is wrong with using
```
PUT /telemetry/{tenantId}/{deviceId}?data=....
```
This is what we defined for publishing data using the gateway mode (which is what the SigFox network provider is, from Hono's point of view).

Reply:
Sure, I can change that. Why do we need the tenant a second time?

Reply:
That's how we defined the URI for the gateway mode. A gateway can act on behalf of a device, which is identified uniquely by its tenant and device ID. We currently only allow gateways to act on behalf of devices of the same tenant as the gateway, but that is just current implementation ...

Reply:
But the tenant is already part of the authentication information of the gateway.

Reply:
What I also miss with this pattern is a way to specify which callback this covers. Sigfox has multiple callbacks, this currently only implement `DATA`, thus the `data` prefix.

Reply:
as I said, we currently verify if the tenant of the authenticated device (the gateway, here: SigFox back end) matches the tenant of the device that the gateway wants to act on behalf of.

Reply:
> What I also miss with this pattern is a way to specify which callback this covers. Sigfox has multiple callbacks, this currently only implement DATA, thus the data prefix.

The data callback seems to map nicely to telemetry, doesn't it?
Which other callbacks need to be handled?

Reply:
DATA maps to telemetry and event.

There are currently 8 callback types. I think the others that may make sense at the moment are: ERROR, SERVICE/STATUS, SERVICE/DATA_ADVANCED.

Reply:
> as I said, the protocol adapters verify that the tenant of the authenticated device (the gateway, here: SigFox back end) matches the tenant of the device that the gateway wants to act on behalf of. However, that may change in the future ...

Ok, but that is already done without the need for a second tenant id (`gatewayDevice.getTenantId()`). Offering a second place to configure the tenant ID will only allow misconfigurations.

Reply:
The tenant that you get with `gatewayDevice.getTenantId()` is the tenant of the device that has authenticated. The tenant in the URI is the tenant of the device that the GW wants to act on behalf of. These are **not necessarily** the same ...

Reply:
I just tried to configure the backend with `PUT`. However this requires to put in a JSON encoded payload, which has to be configuring using a JSON template in the backend. It is not possible to use `application/octet-stream`.

I don't see any benefit in wrapping the while hex encoded data into a JSON object, just to remove that later on.

Could we just stick to the default of using `GET` with the `data` parameter?

Reply:
> The tenant that you get with `gatewayDevice.getTenantId()` is the tenant of the device that has authenticated. The tenant in the URI is the tenant of the device that the GW wants to act on behalf of. These are **not necessarily** the same ...

So just to be sure I understood correctly. You want me to add the tenant into the path, and then fail the message later on, if the tenant of the credentials doesn't match tenant of the URI?!  As currently we don't have any way of granting cross-tenant posting.

Reply:
I just tried to using path parameters, that doesn't work with the backend either. So, the only way use this in a sane and working way is to use GET with query parameters.

Reply:
> So just to be sure I understood correctly. You want me to add the tenant into the path, and then fail the message later on, if the tenant of the credentials doesn't match tenant of the URI?! As currently we don't have any way of granting cross-tenant posting.

I do not want you to do anything. I just wanted to explain the way that the gateway mode works. We use this pattern in all the other protocol adapters. You might even call this a *design principle* :-)

If this protocol adapter doesn't need it, e.g. because it should assume that the device ID provided in the path belongs to the same tenant as the authenticated device, then be my guest to omit it ...

Reply:
> I don't see any benefit in wrapping the while hex encoded data into a JSON object, just to remove that later on.

Neither do I

> Could we just stick to the default of using GET with the data parameter?

Sure

Reply:
Having the tenant in the URI path is also necessary, if the client is not authenticated. A use case that we still support by means of the corresponding configuration property in all protocol adapters.
If the client is not authenticated, then you definitely need to pass in the tenant of the device because you cannot determine it as part of the authentication process.
So, if the SigFox adapter does not support anonymous clients then I think we would not need the tenant in the path. However, in order to provide consistent behavior across all protocol adapters, I think it should ...

Reply:
Ok, I will simply put it in.

Reply:
Ok, the tenant ID is now part of the URL.